core/services/pipeline/runner.go

```diff
@@ -433,30 +444,74 @@ func (r *runner) ExecuteAndInsertFinishedRun(ctx context.Context, spec Spec, var
 		return 0, finalResult, nil
 	}
 
-	if runID, err = r.orm.InsertFinishedRun(r.orm.DB(), run, trrs, saveSuccessfulTaskRuns); err != nil {
+	if runID, err = r.orm.InsertFinishedRun(postgres.UnwrapGormDB(r.orm.DB()), run, saveSuccessfulTaskRuns); err != nil {
 		return runID, finalResult, errors.Wrapf(err, "error inserting finished results for spec ID %v", spec.ID)
 	}
 	return runID, finalResult, nil
 
 }
 
-func (r *runner) Run(ctx context.Context, run *Run, l logger.Logger, saveSuccessfulTaskRuns bool) (incomplete bool, err error) {
+func (r *runner) Run(ctx context.Context, run *Run, l logger.Logger, saveSuccessfulTaskRuns bool, fn func(tx *gorm.DB) error) (incomplete bool, err error) {
+	pipeline, err := r.initializePipeline(run)
+	if err != nil {
+		return false, err
+	}
+
+	async := pipeline.HasAsync()
+
+	err = postgres.GormTransactionWithDefaultContext(r.orm.DB(), func(tx *gorm.DB) error {
+		// avoid an extra db write if there is no async tasks present or if this is a resumed run
+		if async && run.ID == 0 {
```

Comment:
When would the run ID be set? If it was resuming?

Reply:
Yeah, when you're creating a new run you call `pipeline.NewRun` followed by `runner.Run`. On a resume you instead pass in an existing run loaded from the DB.